src/rgw/rgw_lc.cc

```diff
@@ -1442,6 +1436,12 @@ class LCOpAction_Transition : public LCOpAction {
     if (ret < 0) {
       return ret;
     } else {
+      RGWObjState* obj_state{nullptr};
+      ret = obj->get_obj_state(oc.dpp, &obj_state, null_yield, true);
```

Comment:
we don't really need RGWObjState to get attrs, do we? what does `obj->get_attrs()` return at this point?

Reply:
if I am not mistaken, you do indeed need it

Reply:
if you omit get obj state, you'll get an empty attrs sequence

Reply:
`RadosObject::transition_to_cloud()` calls `RadosReadOp::prepare()` to read the head object. this part *should* initialize those attrs:
https://github.com/ceph/ceph/blob/9b6d380/src/rgw/driver/rados/rgw_sal_rados.cc#L2738

Reply:
this is super helpful, but I think we should distinguish between concise fix for the issue (certainly for backport for downstream--and I'd prefer for upstream too), and further rationalization of zipper apis, which we should definitely also do

Reply:
that said, it should I think not be *necessary* to call ReadOp::prepare to get some of the same info, or so I would argue;  goes to refactoring discussion, I guess

Reply:
> `RadosObject::transition_to_cloud()` calls `RadosReadOp::prepare()` to read the head object. this part _should_ initialize those attrs: https://github.com/ceph/ceph/blob/9b6d380/src/rgw/driver/rados/rgw_sal_rados.cc#L2738

`RadosReadOp::prepare()` gets called only in case of `retain_head_object` set to true.

Also @cbodley I noticed set_atomic() is not used while updating head object. Do we need to set it before calling write_meta() in RadosObject::write_tier_object()?

Reply:
> `RadosReadOp::prepare()` gets called only in case of `retain_head_object` set to true.

i see. when `retain_head_object` is false, this later call to `get_obj_state()` would find that exists=false, so the attrs would be empty and we'd call `publish_commit()` with an empty etag

> I noticed set_atomic() is not used while updating head object. Do we need to set it before calling write_meta() in RadosObject::write_tier_object()?

i think so, yes. if another object upload races to overwrite the head, set_atomic() would prevent transition from overwriting that

Reply:
since it's possible for `transition_to_cloud()` to delete the head object, we'd need to read its etag before that. while it's not safe to hold this `RGWObjState*` across this call, we could save the value of its etag

Reply:
In case `retain_head_object` is set to false, the object is deleted after calling `publish_commit()` in `transition_obj_to_cloud`. Since its common to all stores, that code is not moved to RadosObject::transition_to_cloud().

Reply:
I just noticed that in  `remove_expired_obj()` which is used in regular LC expiration too, `publish_commit` is called after the object is deleted - https://github.com/ceph/ceph/blob/9b6d380/src/rgw/rgw_lc.cc#L625 . Is this valid?

Reply:
Will it be cleaner and safe to just save `etag` in all the callers of `publish_commit()` before applying any LC operation (like you mentioned above)?

Reply:
> Will it be cleaner and safe to just save etag in all the callers of publish_commit() before applying any LC operation (like you mentioned above)?

i think so

in the refactoring call, we talked about changes to `get_obj_state()` to avoid the dangling `RGWObjState*`. @dang's suggestion is to rename it to `load_obj_state()` and rely on `StoreObject::state` to store its updated state. so we'd call `obj->load_obj_state()` then read the etag from `obj->get_attrs()`. if you leave the `get_obj_state()` call where it was but just save the etag, it should be easy to follow up with those other changes

what do you think @mattbenjamin @dang?

Reply:
I took the original commit for 7.1, but don't object to the idea, I'll rebase it if we take that version this weekish

Reply:
@dang @cbodley @mattbenjamin updated the PR with the changes as discussed above. Please review.